{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GEN CFTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "FASTA:\n",
      "ID : NM_000492.3\n",
      "Description : NM_000492.3 Homo sapiens CF transmembrane conductance regulator (CFTR), mRNA\n",
      "Sequence Length : 6132\n",
      "Composition : {'A': 1887, 'T': 1731, 'C': 1182, 'G': 1332}\n",
      "\n",
      "AHORA EL FORMATO GENBANK:\n",
      "ID : NM_000492.3\n",
      "Description : Homo sapiens CF transmembrane conductance regulator (CFTR), mRNA\n",
      "Sequence Length : 6132\n",
      "Composition : {'A': 1887, 'T': 1731, 'C': 1182, 'G': 1332}\n",
      "Annotations : {'molecule_type': 'mRNA', 'topology': 'linear', 'data_file_division': 'PRI', 'date': '27-AUG-2019', 'accessions': ['NM_000492'], 'sequence_version': 3, 'keywords': ['RefSeq', 'RefSeq Select'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Domain-interface dynamics of CFTR revealed by stabilizing nanobodies', ...), Reference(title='Unusual Cystic Fibrosis Transmembrane Conductance Regulator Mutations and Liver Disease: A Case Series and Review of the Literature', ...), Reference(title='Spectrum of CFTR mutations in Chechen cystic fibrosis patients: high frequency of c.1545_1546delTA (p.Tyr515X; 1677delTA) and c.274G>A (p.Glu92Lys, E92K) mutations in North Caucasus', ...), Reference(title='Restoration of F508-del Function by Transcomplementation: The Partners Meet in the Endoplasmic Reticulum', ...), Reference(title='Pancreatitis Overview', ...), Reference(title='Cystic Fibrosis and Congenital Absence of the Vas Deferens', ...), Reference(title='A serine to proline substitution (S1255P) in the second nucleotide binding fold of the cystic fibrosis gene', ...), Reference(title='Detection of novel and rare mutations in exon 4 of the cystic fibrosis gene by SSCP', ...), Reference(title='A new missense mutation (R1283M) in exon 20 of the cystic fibrosis transmembrane conductance regulator gene', ...), Reference(title='Three novel mutations in the cystic fibrosis gene detected by chemical cleavage: analysis of variant splicing and a nonsense mutation', ...)], 'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from M28668.1, AC000111.1 and\\nAC000061.1.\\nThis sequence is a reference standard in the RefSeqGene project.\\n[WARNING] On Aug 31, 2019 this sequence was replaced by\\nNM_000492.4.\\nOn Mar 24, 2006 this sequence version replaced NM_000492.2.\\nSummary: This gene encodes a member of the ATP-binding cassette\\n(ABC) transporter superfamily. The encoded protein functions as a\\nchloride channel, making it unique among members of this protein\\nfamily, and controls ion and water secretion and absorption in\\nepithelial tissues. Channel activation is mediated by cycles of\\nregulatory domain phosphorylation, ATP-binding by the\\nnucleotide-binding domains, and ATP hydrolysis. Mutations in this\\ngene cause cystic fibrosis, the most common lethal genetic disorder\\nin populations of Northern European descent. The most frequently\\noccurring mutation in cystic fibrosis, DeltaF508, results in\\nimpaired folding and trafficking of the encoded protein. Multiple\\npseudogenes have been identified in the human genome. [provided by\\nRefSeq, Aug 2017].\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.\\nCOMPLETENESS: full length.', 'structured_comment': defaultdict(<class 'dict'>, {'Evidence-Data': {'Transcript exon combination': 'M28668.1 [ECO:0000332]', 'RNAseq introns': 'single sample supports all introns SAMEA1968540, SAMEA1968832 [ECO:0000348]'}, 'RefSeq-Attributes': {'RefSeq Select criteria': 'based on single protein-coding transcript'}})}\n",
      "Features : [SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(6132), strand=1), type='source', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(6132), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(185), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(66), ExactPosition(69), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(132), ExactPosition(4575), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(363), ExactPosition(426), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(498), ExactPosition(570), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(717), ExactPosition(780), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(798), ExactPosition(861), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1026), ExactPosition(1089), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1149), ExactPosition(1206), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1776), ExactPosition(1779), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2091), ExactPosition(2625), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2109), ExactPosition(2112), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2139), ExactPosition(2142), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2187), ExactPosition(2190), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2229), ExactPosition(2232), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2265), ExactPosition(2268), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2280), ExactPosition(2283), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2340), ExactPosition(2343), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2388), ExactPosition(2391), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2433), ExactPosition(2436), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2499), ExactPosition(2502), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2514), ExactPosition(2517), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2568), ExactPosition(2571), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2706), ExactPosition(2769), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2811), ExactPosition(2814), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2829), ExactPosition(2832), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2886), ExactPosition(2949), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3102), ExactPosition(3165), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3171), ExactPosition(3234), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3417), ExactPosition(3480), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3522), ExactPosition(3585), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4287), ExactPosition(4572), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4461), ExactPosition(4464), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4497), ExactPosition(4500), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4563), ExactPosition(4572), strand=1), type='misc_feature', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(185), ExactPosition(296), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(296), ExactPosition(405), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(405), ExactPosition(621), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(621), ExactPosition(711), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(711), ExactPosition(875), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(875), ExactPosition(1001), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1001), ExactPosition(1248), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1248), ExactPosition(1341), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1341), ExactPosition(1524), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1524), ExactPosition(1716), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1716), ExactPosition(1811), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1811), ExactPosition(1898), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(1898), ExactPosition(2622), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2622), ExactPosition(2751), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2751), ExactPosition(2789), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2789), ExactPosition(3040), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3040), ExactPosition(3120), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3120), ExactPosition(3271), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3271), ExactPosition(3499), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3499), ExactPosition(3600), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3600), ExactPosition(3849), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(3849), ExactPosition(4005), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4005), ExactPosition(4095), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4095), ExactPosition(4268), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4268), ExactPosition(4374), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(4374), ExactPosition(6132), strand=1), type='exon', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(6106), ExactPosition(6112), strand=1), type='regulatory', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(6131), ExactPosition(6132), strand=1), type='polyA_site', qualifiers=...)]\n",
      "Origin : AATTGGAAGCAAATGACATCACAGCAGGTCAGAGAAAAAGGGTTGAGCGGCAGGCACCCAGAGTAGTAGGTCTTTGGCATTAGGAGCTTGAGCCCAGACGGCCCTAGCAGGGACCCCAGCGCCCGAGAGACCATGCAGAGGTCGCCTCTGGAAAAGGCCAGCGTTGTCTCCAAACTTTTTTTCAGCTGGACCAGACCAATTTTGAGGAAAGGATACAGACAGCGCCTGGAATTGTCAGACATATACCAAATCCCTTCTGTTGATTCTGCTGACAATCTATCTGAAAAATTGGAAAGAGAATGGGATAGAGAGCTGGCTTCAAAGAAAAATCCTAAACTCATTAATGCCCTTCGGCGATGTTTTTTCTGGAGATTTATGTTCTATGGAATCTTTTTATATTTAGGGGAAGTCACCAAAGCAGTACAGCCTCTCTTACTGGGAAGAATCATAGCTTCCTATGACCCGGATAACAAGGAGGAACGCTCTATCGCGATTTATCTAGGCATAGGCTTATGCCTTCTCTTTATTGTGAGGACACTGCTCCTACACCCAGCCATTTTTGGCCTTCATCACATTGGAATGCAGATGAGAATAGCTATGTTTAGTTTGATTTATAAGAAGACTTTAAAGCTGTCAAGCCGTGTTCTAGATAAAATAAGTATTGGACAACTTGTTAGTCTCCTTTCCAACAACCTGAACAAATTTGATGAAGGACTTGCATTGGCACATTTCGTGTGGATCGCTCCTTTGCAAGTGGCACTCCTCATGGGGCTAATCTGGGAGTTGTTACAGGCGTCTGCCTTCTGTGGACTTGGTTTCCTGATAGTCCTTGCCCTTTTTCAGGCTGGGCTAGGGAGAATGATGATGAAGTACAGAGATCAGAGAGCTGGGAAGATCAGTGAAAGACTTGTGATTACCTCAGAAATGATTGAAAATATCCAATCTGTTAAGGCATACTGCTGGGAAGAAGCAATGGAAAAAATGATTGAAAACTTAAGACAAACAGAACTGAAACTGACTCGGAAGGCAGCCTATGTGAGATACTTCAATAGCTCAGCCTTCTTCTTCTCAGGGTTCTTTGTGGTGTTTTTATCTGTGCTTCCCTATGCACTAATCAAAGGAATCATCCTCCGGAAAATATTCACCACCATCTCATTCTGCATTGTTCTGCGCATGGCGGTCACTCGGCAATTTCCCTGGGCTGTACAAACATGGTATGACTCTCTTGGAGCAATAAACAAAATACAGGATTTCTTACAAAAGCAAGAATATAAGACATTGGAATATAACTTAACGACTACAGAAGTAGTGATGGAGAATGTAACAGCCTTCTGGGAGGAGGGATTTGGGGAATTATTTGAGAAAGCAAAACAAAACAATAACAATAGAAAAACTTCTAATGGTGATGACAGCCTCTTCTTCAGTAATTTCTCACTTCTTGGTACTCCTGTCCTGAAAGATATTAATTTCAAGATAGAAAGAGGACAGTTGTTGGCGGTTGCTGGATCCACTGGAGCAGGCAAGACTTCACTTCTAATGGTGATTATGGGAGAACTGGAGCCTTCAGAGGGTAAAATTAAGCACAGTGGAAGAATTTCATTCTGTTCTCAGTTTTCCTGGATTATGCCTGGCACCATTAAAGAAAATATCATCTTTGGTGTTTCCTATGATGAATATAGATACAGAAGCGTCATCAAAGCATGCCAACTAGAAGAGGACATCTCCAAGTTTGCAGAGAAAGACAATATAGTTCTTGGAGAAGGTGGAATCACACTGAGTGGAGGTCAACGAGCAAGAATTTCTTTAGCAAGAGCAGTATACAAAGATGCTGATTTGTATTTATTAGACTCTCCTTTTGGATACCTAGATGTTTTAACAGAAAAAGAAATATTTGAAAGCTGTGTCTGTAAACTGATGGCTAACAAAACTAGGATTTTGGTCACTTCTAAAATGGAACATTTAAAGAAAGCTGACAAAATATTAATTTTGCATGAAGGTAGCAGCTATTTTTATGGGACATTTTCAGAACTCCAAAATCTACAGCCAGACTTTAGCTCAAAACTCATGGGATGTGATTCTTTCGACCAATTTAGTGCAGAAAGAAGAAATTCAATCCTAACTGAGACCTTACACCGTTTCTCATTAGAAGGAGATGCTCCTGTCTCCTGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTGGGGAAAAAAGGAAGAATTCTATTCTCAATCCAATCAACTCTATACGAAAATTTTCCATTGTGCAAAAGACTCCCTTACAAATGAATGGCATCGAAGAGGATTCTGATGAGCCTTTAGAGAGAAGGCTGTCCTTAGTACCAGATTCTGAGCAGGGAGAGGCGATACTGCCTCGCATCAGCGTGATCAGCACTGGCCCCACGCTTCAGGCACGAAGGAGGCAGTCTGTCCTGAACCTGATGACACACTCAGTTAACCAAGGTCAGAACATTCACCGAAAGACAACAGCATCCACACGAAAAGTGTCACTGGCCCCTCAGGCAAACTTGACTGAACTGGATATATATTCAAGAAGGTTATCTCAAGAAACTGGCTTGGAAATAAGTGAAGAAATTAACGAAGAAGACTTAAAGGAGTGCTTTTTTGATGATATGGAGAGCATACCAGCAGTGACTACATGGAACACATACCTTCGATATATTACTGTCCACAAGAGCTTAATTTTTGTGCTAATTTGGTGCTTAGTAATTTTTCTGGCAGAGGTGGCTGCTTCTTTGGTTGTGCTGTGGCTCCTTGGAAACACTCCTCTTCAAGACAAAGGGAATAGTACTCATAGTAGAAATAACAGCTATGCAGTGATTATCACCAGCACCAGTTCGTATTATGTGTTTTACATTTACGTGGGAGTAGCCGACACTTTGCTTGCTATGGGATTCTTCAGAGGTCTACCACTGGTGCATACTCTAATCACAGTGTCGAAAATTTTACACCACAAAATGTTACATTCTGTTCTTCAAGCACCTATGTCAACCCTCAACACGTTGAAAGCAGGTGGGATTCTTAATAGATTCTCCAAAGATATAGCAATTTTGGATGACCTTCTGCCTCTTACCATATTTGACTTCATCCAGTTGTTATTAATTGTGATTGGAGCTATAGCAGTTGTCGCAGTTTTACAACCCTACATCTTTGTTGCAACAGTGCCAGTGATAGTGGCTTTTATTATGTTGAGAGCATATTTCCTCCAAACCTCACAGCAACTCAAACAACTGGAATCTGAAGGCAGGAGTCCAATTTTCACTCATCTTGTTACAAGCTTAAAAGGACTATGGACACTTCGTGCCTTCGGACGGCAGCCTTACTTTGAAACTCTGTTCCACAAAGCTCTGAATTTACATACTGCCAACTGGTTCTTGTACCTGTCAACACTGCGCTGGTTCCAAATGAGAATAGAAATGATTTTTGTCATCTTCTTCATTGCTGTTACCTTCATTTCCATTTTAACAACAGGAGAAGGAGAAGGAAGAGTTGGTATTATCCTGACTTTAGCCATGAATATCATGAGTACATTGCAGTGGGCTGTAAACTCCAGCATAGATGTGGATAGCTTGATGCGATCTGTGAGCCGAGTCTTTAAGTTCATTGACATGCCAACAGAAGGTAAACCTACCAAGTCAACCAAACCATACAAGAATGGCCAACTCTCGAAAGTTATGATTATTGAGAATTCACACGTGAAGAAAGATGACATCTGGCCCTCAGGGGGCCAAATGACTGTCAAAGATCTCACAGCAAAATACACAGAAGGTGGAAATGCCATATTAGAGAACATTTCCTTCTCAATAAGTCCTGGCCAGAGGGTGGGCCTCTTGGGAAGAACTGGATCAGGGAAGAGTACTTTGTTATCAGCTTTTTTGAGACTACTGAACACTGAAGGAGAAATCCAGATCGATGGTGTGTCTTGGGATTCAATAACTTTGCAACAGTGGAGGAAAGCCTTTGGAGTGATACCACAGAAAGTATTTATTTTTTCTGGAACATTTAGAAAAAACTTGGATCCCTATGAACAGTGGAGTGATCAAGAAATATGGAAAGTTGCAGATGAGGTTGGGCTCAGATCTGTGATAGAACAGTTTCCTGGGAAGCTTGACTTTGTCCTTGTGGATGGGGGCTGTGTCCTAAGCCATGGCCACAAGCAGTTGATGTGCTTGGCTAGATCTGTTCTCAGTAAGGCGAAGATCTTGCTGCTTGATGAACCCAGTGCTCATTTGGATCCAGTAACATACCAAATAATTAGAAGAACTCTAAAACAAGCATTTGCTGATTGCACAGTAATTCTCTGTGAACACAGGATAGAAGCAATGCTGGAATGCCAACAATTTTTGGTCATAGAAGAGAACAAAGTGCGGCAGTACGATTCCATCCAGAAACTGCTGAACGAGAGGAGCCTCTTCCGGCAAGCCATCAGCCCCTCCGACAGGGTGAAGCTCTTTCCCCACCGGAACTCAAGCAAGTGCAAGTCTAAGCCCCAGATTGCTGCTCTGAAAGAGGAGACAGAAGAAGAGGTGCAAGATACAAGGCTTTAGAGAGCAGCATAAATGTTGACATGGGACATTTGCTCATGGAATTGGAGCTCGTGGGACAGTCACCTCATGGAATTGGAGCTCGTGGAACAGTTACCTCTGCCTCAGAAAACAAGGATGAATTAAGTTTTTTTTTAAAAAAGAAACATTTGGTAAGGGGAATTGAGGACACTGATATGGGTCTTGATAAATGGCTTCCTGGCAATAGTCAAATTGTGTGAAAGGTACTTCAAATCCTTGAAGATTTACCACTTGTGTTTTGCAAGCCAGATTTTCCTGAAAACCCTTGCCATGTGCTAGTAATTGGAAAGGCAGCTCTAAATGTCAATCAGCCTAGTTGATCAGCTTATTGTCTAGTGAAACTCGTTAATTTGTAGTGTTGGAGAAGAACTGAAATCATACTTCTTAGGGTTATGATTAAGTAATGATAACTGGAAACTTCAGCGGTTTATATAAGCTTGTATTCCTTTTTCTCTCCTCTCCCCATGATGTTTAGAAACACAACTATATTGTTTGCTAAGCATTCCAACTATCTCATTTCCAAGCAAGTATTAGAATACCACAGGAACCACAAGACTGCACATCAAAATATGCCCCATTCAACATCTAGTGAGCAGTCAGGAAAGAGAACTTCCAGATCCTGGAAATCAGGGTTAGTATTGTCCAGGTCTACCAAAAATCTCAATATTTCAGATAATCACAATACATCCCTTACCTGGGAAAGGGCTGTTATAATCTTTCACAGGGGACAGGATGGTTCCCTTGATGAAGAAGTTGATATGCCTTTTCCCAACTCCAGAAAGTGACAAGCTCACAGACCTTTGAACTAGAGTTTAGCTGGAAAAGTATGTTAGTGCAAATTGTCACAGGACAGCCCTTCTTTCCACAGAAGCTCCAGGTAGAGGGTGTGTAAGTAGATAGGCCATGGGCACTGTGGGTAGACACACATGAAGTCCAAGCATTTAGATGTATAGGTTGATGGTGGTATGTTTTCAGGCTAGATGTATGTACTTCATGCTGTCTACACTAAGAGAGAATGAGAGACACACTGAAGAAGCACCAATCATGAATTAGTTTTATATGCTTCTGTTTTATAATTTTGTGAAGCAAAATTTTTTCTCTAGGAAATATTTATTTTAATAATGTTTCAAACATATATAACAATGCTGTATTTTAAAAGAATGATTATGAATTACATTTGTATAAAATAATTTTTATATTTGAAATATTGACTTTTTATGGCACTAGTATTTCTATGAAATATTATGTTAAAACTGGGACAGGGGAGAACCTAGGGTGATATTAACCAGGGGCCATGAATCACCTTTTGGTCTGGAGGGAAGCCTTGGGGCTGATGCAGTTGTTGCCCACAGCTGTATGATTCCCAGCCAGCACAGCCTCTTAGATGCAGTTCTGAAGAAGATGGTACCACCAGTCTGACTGTTTCCATCAAGGGTACACTGCCTTCTCAACTCCAAACTGACTCTTAAGAAGACTGCATTATATTTATTACTGTAAGAAAATATCACTTGTCAATAAAATCCATACATTTGTGTGAAA\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Paths to the uploaded files\n",
    "fasta_file_path = 'genes/CFTR/NM_000492.3.fasta'\n",
    "genbank_file_path = 'genes/CFTR/NM_000492.3.gb'\n",
    "\n",
    "# Parse the files to extract sequence information\n",
    "fasta_seq = SeqIO.read(fasta_file_path, \"fasta\")\n",
    "genbank_seq = SeqIO.read(genbank_file_path, \"genbank\")\n",
    "\n",
    "# Extract relevant information\n",
    "fasta_info = {\n",
    "    \"ID\": fasta_seq.id,\n",
    "    \"Description\": fasta_seq.description,\n",
    "    \"Sequence Length\": len(fasta_seq.seq),\n",
    "    \"Composition\": {\n",
    "        \"A\": fasta_seq.seq.count(\"A\"),\n",
    "        \"T\": fasta_seq.seq.count(\"T\"),\n",
    "        \"C\": fasta_seq.seq.count(\"C\"),\n",
    "        \"G\": fasta_seq.seq.count(\"G\")\n",
    "    }\n",
    "}\n",
    "\n",
    "genbank_info = {\n",
    "    \"ID\": genbank_seq.id,\n",
    "    \"Description\": genbank_seq.description,\n",
    "    \"Sequence Length\": len(genbank_seq.seq),\n",
    "    \"Composition\": {\n",
    "        \"A\": genbank_seq.seq.count(\"A\"),\n",
    "        \"T\": genbank_seq.seq.count(\"T\"),\n",
    "        \"C\": genbank_seq.seq.count(\"C\"),\n",
    "        \"G\": genbank_seq.seq.count(\"G\")\n",
    "    },\n",
    "    \"Annotations\": genbank_seq.annotations,\n",
    "    \"Features\": genbank_seq.features,\n",
    "    \"Origin\": genbank_seq.seq    \n",
    "\n",
    "}\n",
    "\n",
    "#itera sobre los diccionarios para imprimir la informacion de genbank\n",
    "print(\"\\nFASTA:\")\n",
    "for key, value in fasta_info.items():\n",
    "    print(key, \":\", value)\n",
    "\n",
    "print(\"\\nAHORA EL FORMATO GENBANK:\")\n",
    "for key, value in genbank_info.items():\n",
    "    print(key, \":\", value)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Análisis de la Información de los Archivos FASTA y GenBank\n",
    "\n",
    "La salida obtenida nos presenta información sobre una secuencia genética del gen **CFTR** (*Cystic Fibrosis Transmembrane Conductance Regulator*), clave en el estudio de enfermedades genéticas como la fibrosis quística. Este análisis se basa en dos formatos diferentes: FASTA y GenBank, ambos ampliamente utilizados en bioinformática.\n",
    "\n",
    "#### Datos en Formato FASTA\n",
    "\n",
    "El archivo FASTA proporciona una descripción básica y la composición de la secuencia. En este caso, la secuencia tiene el ID `NM_000492.3` y una longitud total de 6132 nucleótidos. Su descripción menciona que corresponde al mRNA del gen CFTR en humanos (*Homo sapiens*). En cuanto a su composición, está formada por:\n",
    "\n",
    "- **Adenina (A)**: 1887 nucleótidos\n",
    "- **Timina (T)**: 1731 nucleótidos\n",
    "- **Citosina (C)**: 1182 nucleótidos\n",
    "- **Guanina (G)**: 1332 nucleótidos\n",
    "\n",
    "Este formato es útil para análisis rápidos, ya que resume la secuencia y su composición básica, sin profundizar en anotaciones u otras características. La secuencia proporcionada corresponde a un gen específico, el **CFTR**, relacionado con la fibrosis quística.\n",
    "\n",
    "#### Datos en Formato GenBank\n",
    "\n",
    "Por otro lado, el archivo GenBank incluye una mayor cantidad de información y anotaciones detalladas. Además de la descripción y composición ya mencionadas, se destacan los siguientes puntos:\n",
    "\n",
    "- **Anotaciones generales**:\n",
    "  - Tipo de molécula: mRNA\n",
    "  - Topología: lineal\n",
    "  - División de archivo: PRI (Primates)\n",
    "  - Fecha: 27 de agosto de 2019\n",
    "- **Accesiones y versión**: El archivo corresponde a la versión 3 de la secuencia (`NM_000492.3`), aunque incluye una nota indicando que fue reemplazada posteriormente por la versión 4.\n",
    "- **Palabras clave**: Se destacan términos como *RefSeq* y *RefSeq Select*, que indican que esta secuencia ha sido revisada por el NCBI.\n",
    "- **Fuente y taxonomía**: Se identifica como *Homo sapiens*, con una clasificación detallada que incluye hasta el nivel de especie.\n",
    "- **Referencias bibliográficas**: Contiene múltiples referencias científicas relacionadas con el gen CFTR, lo que indica que esta secuencia es de gran relevancia en investigaciones.\n",
    "- **Features (Características)**: El archivo GenBank proporciona un conjunto detallado de *features* que describen elementos importantes de la secuencia, como:\n",
    "  - **source**: Describe el origen de la secuencia, que se extiende desde la posición 0 hasta la 6132, indicando que es una secuencia completa.\n",
    "  - **gene**: Define la región del gen CFTR.\n",
    "  - **exon**: Incluye múltiples exones que abarcan diferentes posiciones a lo largo de la secuencia, indicando las regiones que se transcriben y se mantienen tras el procesamiento del RNA.\n",
    "  - **misc\\_feature** y **CDS**: Se incluyen diversas *features* que destacan regiones de interés específicas, como sitios funcionales, dominios proteicos, y la región de codificación de la proteína (CDS).\n",
    "  - **regulatory** y **polyA\\_site**: Se mencionan sitios regulatorios y de poliadenilación que son importantes para la expresión y estabilidad del mRNA.\n",
    "\n",
    "El archivo también incluye comentarios estructurados que describen detalles técnicos adicionales:\n",
    "\n",
    "- **Estructura del Gen**: Incluye información sobre la combinación de exones del transcripto (`M28668.1 [ECO:0000332]`) y soporte basado en RNA-seq para los intrones, con ejemplos como `SAMEA1968540` y `SAMEA1968832` que respaldan todos los intrones observados (`[ECO:0000348]`).\n",
    "- **Criterios de RefSeq**: Se menciona que los atributos de *RefSeq* están basados en un transcripto codificante de una sola proteína, lo cual indica que esta secuencia es una referencia importante y seleccionada cuidadosamente.\n",
    "\n",
    "**Origen de la Secuencia (Origin)**: La sección *Origin* del archivo GenBank proporciona la secuencia de nucleótidos completa del gen CFTR, organizada en bloques de 60 bases para facilitar la lectura. Esta secuencia permite a los investigadores analizar en detalle cada región, identificar mutaciones específicas y entender la composición de la secuencia a nivel de bases.\n",
    "\n",
    "Además, se proporciona un resumen funcional del gen CFTR, destacando su papel en la regulación del transporte de iones y su relación con mutaciones que causan fibrosis quística."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PROTEINA 4INS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "FASTA:\n",
      "Sequence ID: pdb|4INS|D\n",
      "Description: pdb|4INS|D Chain D, INSULIN (CHAIN B)\n",
      "Sequence Length: 30\n",
      "Composition: {'A': 2, 'C': 2, 'D': 0, 'E': 2, 'F': 3, 'G': 3, 'H': 2, 'I': 0, 'K': 1, 'L': 4, 'M': 0, 'N': 1, 'P': 1, 'Q': 1, 'R': 1, 'S': 1, 'T': 1, 'V': 3, 'W': 0, 'Y': 2}\n",
      "\n",
      "\n",
      "\n",
      "FASTA:\n",
      "Sequence ID: pdb|4INS|C\n",
      "Description: pdb|4INS|C Chain C, INSULIN (CHAIN A)\n",
      "Sequence Length: 21\n",
      "Composition: {'A': 0, 'C': 4, 'D': 0, 'E': 2, 'F': 0, 'G': 1, 'H': 0, 'I': 2, 'K': 0, 'L': 2, 'M': 0, 'N': 2, 'P': 0, 'Q': 2, 'R': 0, 'S': 2, 'T': 1, 'V': 1, 'W': 0, 'Y': 2}\n",
      "\n",
      "\n",
      "\n",
      "FASTA:\n",
      "Sequence ID: pdb|4INS|B\n",
      "Description: pdb|4INS|B Chain B, INSULIN (CHAIN B)\n",
      "Sequence Length: 30\n",
      "Composition: {'A': 2, 'C': 2, 'D': 0, 'E': 2, 'F': 3, 'G': 3, 'H': 2, 'I': 0, 'K': 1, 'L': 4, 'M': 0, 'N': 1, 'P': 1, 'Q': 1, 'R': 1, 'S': 1, 'T': 1, 'V': 3, 'W': 0, 'Y': 2}\n",
      "\n",
      "\n",
      "\n",
      "FASTA:\n",
      "Sequence ID: pdb|4INS|A\n",
      "Description: pdb|4INS|A Chain A, INSULIN (CHAIN A)\n",
      "Sequence Length: 21\n",
      "Composition: {'A': 0, 'C': 4, 'D': 0, 'E': 2, 'F': 0, 'G': 1, 'H': 0, 'I': 2, 'K': 0, 'L': 2, 'M': 0, 'N': 2, 'P': 0, 'Q': 2, 'R': 0, 'S': 2, 'T': 1, 'V': 1, 'W': 0, 'Y': 2}\n",
      "\n",
      "\n",
      "\n",
      "AHORA EL FORMATO GENPEPT:\n",
      "Sequence ID: 4INS_D\n",
      "Description: Chain D, INSULIN (CHAIN B)\n",
      "Sequence Length: 30\n",
      "Amino Acid Composition: {'A': 2, 'C': 2, 'D': 0, 'E': 2, 'F': 3, 'G': 3, 'H': 2, 'I': 0, 'K': 1, 'L': 4, 'M': 0, 'N': 1, 'P': 1, 'Q': 1, 'R': 1, 'S': 1, 'T': 1, 'V': 3, 'W': 0, 'Y': 2}\n",
      "Annotations: {'topology': 'linear', 'data_file_division': 'MAM', 'date': '23-OCT-2024', 'accessions': ['4INS_D'], 'db_source': 'pdb: molecule 4INS, chain D, release Oct 23, 2024; deposition: Jul 10, 1989; class: HORMONE; source: Mmdb_id: 3182, Pdb_id 1: 4INS; Exp. method: X-ray Diffraction.', 'keywords': [''], 'source': 'Sus scrofa (pig)', 'organism': 'Sus scrofa', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Artiodactyla', 'Suina', 'Suidae', 'Sus'], 'references': [Reference(title='Structure of Rhombohedral 2 Zinc Insulin Crystals', ...), Reference(title='X-Ray Diffraction Data on Some Crystalline Varieties of Insulin', ...), Reference(title='Atomic Positions in Rhombohedral 2-Zinc Insulin Crystals', ...), Reference(title='X-Ray Analysis and the Structure of Insulin', ...), Reference(title='Insulin. The Structure in the Crystal and its Reflection in Chemistry and Biology', ...), Reference(title='The Crystal Structure of Rhombohedral 2 Zinc Insulin', ...), Reference(title='The Structure of Insulin', ...), Reference(title='Varieties of Insulin', ...), Reference(title='A Method for Fitting Satisfactory Models to Sets of Atomic Positions in Protein Structure Refinements', ...), Reference(title='Rhombohedral Insulin Crystal Transformation', ...), Reference(title='Experience with Fast Fourier Least Squares in the Refinement of the Crystal Structure of Rhombohedral 2-Zinc Insulin at 1.5 Angstroms Resolution', ...), Reference(title='Structural Relationships in the Two-Zinc Insulin Hexamer', ...), Reference(title='A Comparative Assessment of the Zinc-Protein Coordination in 2Zn-Insulin as Determined by X-Ray Absorption Fine Structure (Exafs) and X-Ray Crystallography', ...), Reference(title='The structure of 2Zn pig insulin crystals at 1.5 A resolution', ...), Reference(title='Direct Submission', ...)], 'comment': 'THE STRUCTURE OF 2ZN PIG INSULIN CRYSTALS AT 1.5 ANGSTROMS\\nRESOLUTION.', 'molecule_type': 'protein'}\n",
      "Features: [SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(30)), type='source', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(7), ExactPosition(19)), type='SecStr', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(9), ExactPosition(10)), type='Het', qualifiers=...)]\n",
      "Orgin: FVNQHLCGSHLVEALYLVCGERGFFYTPKA\n",
      "\n",
      "\n",
      "\n",
      "AHORA EL FORMATO GENPEPT:\n",
      "Sequence ID: 4INS_C\n",
      "Description: Chain C, INSULIN (CHAIN A)\n",
      "Sequence Length: 21\n",
      "Amino Acid Composition: {'A': 0, 'C': 4, 'D': 0, 'E': 2, 'F': 0, 'G': 1, 'H': 0, 'I': 2, 'K': 0, 'L': 2, 'M': 0, 'N': 2, 'P': 0, 'Q': 2, 'R': 0, 'S': 2, 'T': 1, 'V': 1, 'W': 0, 'Y': 2}\n",
      "Annotations: {'topology': 'linear', 'data_file_division': 'MAM', 'date': '23-OCT-2024', 'accessions': ['4INS_C'], 'db_source': 'pdb: molecule 4INS, chain C, release Oct 23, 2024; deposition: Jul 10, 1989; class: HORMONE; source: Mmdb_id: 3182, Pdb_id 1: 4INS; Exp. method: X-ray Diffraction.', 'keywords': [''], 'source': 'Sus scrofa (pig)', 'organism': 'Sus scrofa', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Artiodactyla', 'Suina', 'Suidae', 'Sus'], 'references': [Reference(title='Structure of Rhombohedral 2 Zinc Insulin Crystals', ...), Reference(title='X-Ray Diffraction Data on Some Crystalline Varieties of Insulin', ...), Reference(title='Atomic Positions in Rhombohedral 2-Zinc Insulin Crystals', ...), Reference(title='X-Ray Analysis and the Structure of Insulin', ...), Reference(title='Insulin. The Structure in the Crystal and its Reflection in Chemistry and Biology', ...), Reference(title='The Crystal Structure of Rhombohedral 2 Zinc Insulin', ...), Reference(title='The Structure of Insulin', ...), Reference(title='Varieties of Insulin', ...), Reference(title='A Method for Fitting Satisfactory Models to Sets of Atomic Positions in Protein Structure Refinements', ...), Reference(title='Rhombohedral Insulin Crystal Transformation', ...), Reference(title='Experience with Fast Fourier Least Squares in the Refinement of the Crystal Structure of Rhombohedral 2-Zinc Insulin at 1.5 Angstroms Resolution', ...), Reference(title='Structural Relationships in the Two-Zinc Insulin Hexamer', ...), Reference(title='A Comparative Assessment of the Zinc-Protein Coordination in 2Zn-Insulin as Determined by X-Ray Absorption Fine Structure (Exafs) and X-Ray Crystallography', ...), Reference(title='The structure of 2Zn pig insulin crystals at 1.5 A resolution', ...), Reference(title='Direct Submission', ...)], 'comment': 'THE STRUCTURE OF 2ZN PIG INSULIN CRYSTALS AT 1.5 ANGSTROMS\\nRESOLUTION.', 'molecule_type': 'protein'}\n",
      "Features: [SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(21)), type='source', qualifiers=...), SeqFeature(CompoundLocation([SimpleLocation(ExactPosition(5), ExactPosition(6)), SimpleLocation(ExactPosition(10), ExactPosition(11))], 'bond'), type='Bond', qualifiers=...)]\n",
      "Orgin: GIVEQCCTSICSLYQLENYCN\n",
      "\n",
      "\n",
      "\n",
      "AHORA EL FORMATO GENPEPT:\n",
      "Sequence ID: 4INS_B\n",
      "Description: Chain B, INSULIN (CHAIN B)\n",
      "Sequence Length: 30\n",
      "Amino Acid Composition: {'A': 2, 'C': 2, 'D': 0, 'E': 2, 'F': 3, 'G': 3, 'H': 2, 'I': 0, 'K': 1, 'L': 4, 'M': 0, 'N': 1, 'P': 1, 'Q': 1, 'R': 1, 'S': 1, 'T': 1, 'V': 3, 'W': 0, 'Y': 2}\n",
      "Annotations: {'topology': 'linear', 'data_file_division': 'MAM', 'date': '23-OCT-2024', 'accessions': ['4INS_B'], 'db_source': 'pdb: molecule 4INS, chain B, release Oct 23, 2024; deposition: Jul 10, 1989; class: HORMONE; source: Mmdb_id: 3182, Pdb_id 1: 4INS; Exp. method: X-ray Diffraction.', 'keywords': [''], 'source': 'Sus scrofa (pig)', 'organism': 'Sus scrofa', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Artiodactyla', 'Suina', 'Suidae', 'Sus'], 'references': [Reference(title='Structure of Rhombohedral 2 Zinc Insulin Crystals', ...), Reference(title='X-Ray Diffraction Data on Some Crystalline Varieties of Insulin', ...), Reference(title='Atomic Positions in Rhombohedral 2-Zinc Insulin Crystals', ...), Reference(title='X-Ray Analysis and the Structure of Insulin', ...), Reference(title='Insulin. The Structure in the Crystal and its Reflection in Chemistry and Biology', ...), Reference(title='The Crystal Structure of Rhombohedral 2 Zinc Insulin', ...), Reference(title='The Structure of Insulin', ...), Reference(title='Varieties of Insulin', ...), Reference(title='A Method for Fitting Satisfactory Models to Sets of Atomic Positions in Protein Structure Refinements', ...), Reference(title='Rhombohedral Insulin Crystal Transformation', ...), Reference(title='Experience with Fast Fourier Least Squares in the Refinement of the Crystal Structure of Rhombohedral 2-Zinc Insulin at 1.5 Angstroms Resolution', ...), Reference(title='Structural Relationships in the Two-Zinc Insulin Hexamer', ...), Reference(title='A Comparative Assessment of the Zinc-Protein Coordination in 2Zn-Insulin as Determined by X-Ray Absorption Fine Structure (Exafs) and X-Ray Crystallography', ...), Reference(title='The structure of 2Zn pig insulin crystals at 1.5 A resolution', ...), Reference(title='Direct Submission', ...)], 'comment': 'THE STRUCTURE OF 2ZN PIG INSULIN CRYSTALS AT 1.5 ANGSTROMS\\nRESOLUTION.', 'molecule_type': 'protein'}\n",
      "Features: [SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(30)), type='source', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(9), ExactPosition(19)), type='SecStr', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(9), ExactPosition(10)), type='Het', qualifiers=...)]\n",
      "Orgin: FVNQHLCGSHLVEALYLVCGERGFFYTPKA\n",
      "\n",
      "\n",
      "\n",
      "AHORA EL FORMATO GENPEPT:\n",
      "Sequence ID: 4INS_A\n",
      "Description: Chain A, INSULIN (CHAIN A)\n",
      "Sequence Length: 21\n",
      "Amino Acid Composition: {'A': 0, 'C': 4, 'D': 0, 'E': 2, 'F': 0, 'G': 1, 'H': 0, 'I': 2, 'K': 0, 'L': 2, 'M': 0, 'N': 2, 'P': 0, 'Q': 2, 'R': 0, 'S': 2, 'T': 1, 'V': 1, 'W': 0, 'Y': 2}\n",
      "Annotations: {'topology': 'linear', 'data_file_division': 'MAM', 'date': '23-OCT-2024', 'accessions': ['4INS_A'], 'db_source': 'pdb: molecule 4INS, chain A, release Oct 23, 2024; deposition: Jul 10, 1989; class: HORMONE; source: Mmdb_id: 3182, Pdb_id 1: 4INS; Exp. method: X-ray Diffraction.', 'keywords': [''], 'source': 'Sus scrofa (pig)', 'organism': 'Sus scrofa', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Artiodactyla', 'Suina', 'Suidae', 'Sus'], 'references': [Reference(title='Structure of Rhombohedral 2 Zinc Insulin Crystals', ...), Reference(title='X-Ray Diffraction Data on Some Crystalline Varieties of Insulin', ...), Reference(title='Atomic Positions in Rhombohedral 2-Zinc Insulin Crystals', ...), Reference(title='X-Ray Analysis and the Structure of Insulin', ...), Reference(title='Insulin. The Structure in the Crystal and its Reflection in Chemistry and Biology', ...), Reference(title='The Crystal Structure of Rhombohedral 2 Zinc Insulin', ...), Reference(title='The Structure of Insulin', ...), Reference(title='Varieties of Insulin', ...), Reference(title='A Method for Fitting Satisfactory Models to Sets of Atomic Positions in Protein Structure Refinements', ...), Reference(title='Rhombohedral Insulin Crystal Transformation', ...), Reference(title='Experience with Fast Fourier Least Squares in the Refinement of the Crystal Structure of Rhombohedral 2-Zinc Insulin at 1.5 Angstroms Resolution', ...), Reference(title='Structural Relationships in the Two-Zinc Insulin Hexamer', ...), Reference(title='A Comparative Assessment of the Zinc-Protein Coordination in 2Zn-Insulin as Determined by X-Ray Absorption Fine Structure (Exafs) and X-Ray Crystallography', ...), Reference(title='The structure of 2Zn pig insulin crystals at 1.5 A resolution', ...), Reference(title='Direct Submission', ...)], 'comment': 'THE STRUCTURE OF 2ZN PIG INSULIN CRYSTALS AT 1.5 ANGSTROMS\\nRESOLUTION.', 'molecule_type': 'protein'}\n",
      "Features: [SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(21)), type='source', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(9)), type='SecStr', qualifiers=...), SeqFeature(CompoundLocation([SimpleLocation(ExactPosition(5), ExactPosition(6)), SimpleLocation(ExactPosition(10), ExactPosition(11))], 'bond'), type='Bond', qualifiers=...)]\n",
      "Orgin: GIVEQCCTSICSLYQLENYCN\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Path to the uploaded FASTA file\n",
    "fasta_file_path = 'proteinas/4INS/4INS.fasta'\n",
    "genpept_file_path = 'proteinas/4INS/4INS.gp'  # Asegúrate de tener el archivo en formato GenPept\n",
    "\n",
    "\n",
    "# Parse the FASTA file to extract multiple sequences\n",
    "fasta_sequences = list(SeqIO.parse(fasta_file_path, \"fasta\"))\n",
    "genpept_sequences = list(SeqIO.parse(genpept_file_path, \"genbank\"))  # GenPept usa el mismo parser que GenBank\n",
    "\n",
    "# Iterate over each sequence and extract relevant information\n",
    "for seq_record in fasta_sequences:\n",
    "    fasta_info = {\n",
    "        \"ID\": seq_record.id,\n",
    "        \"Description\": seq_record.description,\n",
    "        \"Sequence Length\": len(seq_record.seq),\n",
    "        \"Composition\": {\n",
    "            \"A\": seq_record.seq.count(\"A\"),\n",
    "            \"C\": seq_record.seq.count(\"C\"),\n",
    "            \"D\": seq_record.seq.count(\"D\"),\n",
    "            \"E\": seq_record.seq.count(\"E\"),\n",
    "            \"F\": seq_record.seq.count(\"F\"),\n",
    "            \"G\": seq_record.seq.count(\"G\"),\n",
    "            \"H\": seq_record.seq.count(\"H\"),\n",
    "            \"I\": seq_record.seq.count(\"I\"),\n",
    "            \"K\": seq_record.seq.count(\"K\"),\n",
    "            \"L\": seq_record.seq.count(\"L\"),\n",
    "            \"M\": seq_record.seq.count(\"M\"),\n",
    "            \"N\": seq_record.seq.count(\"N\"),\n",
    "            \"P\": seq_record.seq.count(\"P\"),\n",
    "            \"Q\": seq_record.seq.count(\"Q\"),\n",
    "            \"R\": seq_record.seq.count(\"R\"),\n",
    "            \"S\": seq_record.seq.count(\"S\"),\n",
    "            \"T\": seq_record.seq.count(\"T\"),\n",
    "            \"V\": seq_record.seq.count(\"V\"),\n",
    "            \"W\": seq_record.seq.count(\"W\"),\n",
    "            \"Y\": seq_record.seq.count(\"Y\")\n",
    "        }\n",
    "    }\n",
    "\n",
    "    print(\"\\nFASTA:\")\n",
    "    # Print the extracted information for each sequence\n",
    "    print(f\"Sequence ID: {fasta_info['ID']}\")\n",
    "    print(f\"Description: {fasta_info['Description']}\")\n",
    "    print(f\"Sequence Length: {fasta_info['Sequence Length']}\")\n",
    "    print(\"Composition:\", fasta_info[\"Composition\"])\n",
    "    print(\"\\n\")\n",
    "\n",
    "# Iterate over the GenPept sequences and extract relevant information\n",
    "for seq_record in genpept_sequences:\n",
    "    genpept_info = {\n",
    "        \"ID\": seq_record.id,\n",
    "        \"Description\": seq_record.description,\n",
    "        \"Sequence Length\": len(seq_record.seq),\n",
    "        \"Amino Acid Composition\": {\n",
    "            \"A\": seq_record.seq.count(\"A\"),\n",
    "            \"C\": seq_record.seq.count(\"C\"),\n",
    "            \"D\": seq_record.seq.count(\"D\"),\n",
    "            \"E\": seq_record.seq.count(\"E\"),\n",
    "            \"F\": seq_record.seq.count(\"F\"),\n",
    "            \"G\": seq_record.seq.count(\"G\"),\n",
    "            \"H\": seq_record.seq.count(\"H\"),\n",
    "            \"I\": seq_record.seq.count(\"I\"),\n",
    "            \"K\": seq_record.seq.count(\"K\"),\n",
    "            \"L\": seq_record.seq.count(\"L\"),\n",
    "            \"M\": seq_record.seq.count(\"M\"),\n",
    "            \"N\": seq_record.seq.count(\"N\"),\n",
    "            \"P\": seq_record.seq.count(\"P\"),\n",
    "            \"Q\": seq_record.seq.count(\"Q\"),\n",
    "            \"R\": seq_record.seq.count(\"R\"),\n",
    "            \"S\": seq_record.seq.count(\"S\"),\n",
    "            \"T\": seq_record.seq.count(\"T\"),\n",
    "            \"V\": seq_record.seq.count(\"V\"),\n",
    "            \"W\": seq_record.seq.count(\"W\"),\n",
    "            \"Y\": seq_record.seq.count(\"Y\"),\n",
    "        },\n",
    "        \"Annotations\": seq_record.annotations,\n",
    "        \"Features\": seq_record.features,\n",
    "        \"Origin\": seq_record.seq\n",
    "    }\n",
    "\n",
    "    print(\"\\nAHORA EL FORMATO GENPEPT:\")\n",
    "    # Print the extracted information for each protein sequence\n",
    "    print(f\"Sequence ID: {genpept_info['ID']}\")\n",
    "    print(f\"Description: {genpept_info['Description']}\")\n",
    "    print(f\"Sequence Length: {genpept_info['Sequence Length']}\")\n",
    "    print(\"Amino Acid Composition:\", genpept_info[\"Amino Acid Composition\"])\n",
    "    print(\"Annotations:\", genpept_info[\"Annotations\"])\n",
    "    print(\"Features:\", genpept_info[\"Features\"])\n",
    "    print(\"Orgin:\", genpept_info[\"Origin\"])\n",
    "    print(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Análisis de la Información de los Archivos FASTA y GenPept para la Proteína 4INS\n",
    "\n",
    "La salida obtenida nos presenta información sobre la estructura proteica de la **insulina** derivada de **cerdo (Sus scrofa)**, con el identificador **4INS**. Este análisis se basa en dos formatos diferentes: **FASTA** y **GenPept**, ambos ampliamente utilizados en bioinformática.\n",
    "\n",
    "#### Datos en Formato FASTA\n",
    "El archivo FASTA proporciona una descripción básica y la composición de las cadenas proteicas de la insulina. En este caso, las secuencias corresponden a cuatro cadenas diferentes de la proteína insulina:\n",
    "\n",
    "- **Cadena D (INSULIN, CHAIN B)**: \n",
    "  - **Secuencia de aminoácidos**: `FVNQHLCGSHLVEALYLVCGERGFFYTPKA`\n",
    "  - **Longitud de la secuencia**: 30 aminoácidos\n",
    "  - **Composición**: A: 2, C: 3, D: 0, E: 2, F: 3, G: 3, H: 1, I: 0, K: 1, L: 4, M: 0, N: 1, P: 1, Q: 1, R: 1, S: 1, T: 1, V: 3, W: 0, Y: 2\n",
    "\n",
    "- **Cadena C (INSULIN, CHAIN A)**: \n",
    "  - **Secuencia de aminoácidos**: `GIVEQCCTSICSLYQLENYCN`\n",
    "  - **Longitud de la secuencia**: 21 aminoácidos\n",
    "  - **Composición**: A: 0, C: 4, D: 0, E: 2, F: 0, G: 1, H: 0, I: 2, K: 0, L: 2, M: 0, N: 2, P: 0, Q: 2, R: 0, S: 2, T: 1, V: 1, W: 0, Y: 2\n",
    "\n",
    "- **Cadena B (INSULIN, CHAIN B)**: \n",
    "  - **Secuencia de aminoácidos**: `FVNQHLCGSHLVEALYLVCGERGFFYTPKA`\n",
    "  - **Longitud de la secuencia**: 30 aminoácidos\n",
    "  - **Composición**: A: 2, C: 3, D: 0, E: 2, F: 3, G: 3, H: 1, I: 0, K: 1, L: 4, M: 0, N: 1, P: 1, Q: 1, R: 1, S: 1, T: 1, V: 3, W: 0, Y: 2\n",
    "\n",
    "- **Cadena A (INSULIN, CHAIN A)**: \n",
    "  - **Secuencia de aminoácidos**: `GIVEQCCTSICSLYQLENYCN`\n",
    "  - **Longitud de la secuencia**: 21 aminoácidos\n",
    "  - **Composición**: A: 0, C: 4, D: 0, E: 2, F: 0, G: 1, H: 0, I: 2, K: 0, L: 2, M: 0, N: 2, P: 0, Q: 2, R: 0, S: 2, T: 1, V: 1, W: 0, Y: 2\n",
    "\n",
    "Este formato es útil para un análisis rápido de la secuencia de la proteína y su composición básica.\n",
    "\n",
    "#### Datos en Formato GenPept\n",
    "El archivo GenPept incluye una mayor cantidad de información y anotaciones detalladas de la estructura de la insulina. Se destacan los siguientes puntos para cada una de las cadenas:\n",
    "\n",
    "- **Cadena D (INSULIN, CHAIN B)**:\n",
    "  - **Secuencia de aminoácidos**: `FVNQHLCGSHLVEALYLVCGERGFFYTPKA`\n",
    "  - **Longitud de la secuencia**: 30 aminoácidos\n",
    "  - **Composición de aminoácidos**: A: 2, C: 3, D: 0, E: 2, F: 3, G: 3, H: 1, I: 0, K: 1, L: 4, M: 0, N: 1, P: 1, Q: 1, R: 1, S: 1, T: 1, V: 3, W: 0, Y: 2\n",
    "  - **Anotaciones**: La cadena pertenece al organismo **Sus scrofa** (cerdo), con método experimental de **difracción de rayos X**. Se destaca que el origen de la secuencia es de **insulina de cerdo**, y se incluyen referencias sobre su estructura cristalina.\n",
    "\n",
    "- **Cadena C (INSULIN, CHAIN A)**:\n",
    "  - **Secuencia de aminoácidos**: `GIVEQCCTSICSLYQLENYCN`\n",
    "  - **Longitud de la secuencia**: 21 aminoácidos\n",
    "  - **Composición de aminoácidos**: A: 0, C: 4, D: 0, E: 2, F: 0, G: 1, H: 0, I: 2, K: 0, L: 2, M: 0, N: 2, P: 0, Q: 2, R: 0, S: 2, T: 1, V: 1, W: 0, Y: 2\n",
    "  - **Anotaciones**: Similarmente, la cadena se deriva de **Sus scrofa** y cuenta con anotaciones detalladas sobre la estructura y referencias bibliográficas sobre su estudio.\n",
    "\n",
    "- **Cadena B (INSULIN, CHAIN B)**:\n",
    "  - **Secuencia de aminoácidos**: `FVNQHLCGSHLVEALYLVCGERGFFYTPKA`\n",
    "  - **Longitud de la secuencia**: 30 aminoácidos\n",
    "  - **Composición de aminoácidos**: A: 2, C: 3, D: 0, E: 2, F: 3, G: 3, H: 1, I: 0, K: 1, L: 4, M: 0, N: 1, P: 1, Q: 1, R: 1, S: 1, T: 1, V: 3, W: 0, Y: 2\n",
    "  - **Anotaciones**: La información es similar a la cadena D, describiendo la estructura y referencias de la insulina.\n",
    "\n",
    "- **Cadena A (INSULIN, CHAIN A)**:\n",
    "  - **Secuencia de aminoácidos**: `GIVEQCCTSICSLYQLENYCN`\n",
    "  - **Longitud de la secuencia**: 21 aminoácidos\n",
    "  - **Composición de aminoácidos**: A: 0, C: 4, D: 0, E: 2, F: 0, G: 1, H: 0, I: 2, K: 0, L: 2, M: 0, N: 2, P: 0, Q: 2, R: 0, S: 2, T: 1, V: 1, W: 0, Y: 2\n",
    "  - **Anotaciones**: Esta cadena también pertenece a **Sus scrofa** y proporciona detalles adicionales sobre la estructura proteica.\n",
    "\n",
    "En el formato GenPept, las anotaciones incluyen información relevante sobre el origen de la proteína, las técnicas experimentales utilizadas para su determinación (como difracción de rayos X), así como referencias bibliográficas detalladas. Esto permite un análisis más profundo de la estructura y funcionalidad de la insulina en comparación con el archivo FASTA."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusión de ambos formatos:\n",
    "FASTA es ideal para obtener una visión rápida de la secuencia de aminoácidos, mientras que GenPept nos brinda una visión detallada de la proteína o el gen, incluyendo contexto biológico y experimental. FASTA es rápido y directo; GenPept es completo y profundo. Ambos juntos nos permiten entender la insulina desde sus componentes básicos hasta su importancia biológica y estructural.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
